JS/Lesson19/goksel/recursion.js

```diff
@@ -36,103 +36,231 @@ console.log('5', getFactorial(5));
+        const firstWordCapitalized = firstWord[0].toUpperCase() + firstWord.slice(1);
+        const restOfTheSentenceArray = sentenceArray.slice(1);
+        const restOfTheSentence = restOfTheSentenceArray.join(' ');
+        return firstWordCapitalized + ' ' + capitalizeFirstLetter(restOfTheSentence) ;
+    } 
 }
 
 console.log(capitalizeFirstLetter(ex2sentence));
 
 
+
+console.log("---Ex3---");
 // 3. Write a recursive function to count the number of vowels in a string.
 const sentenceEx4 = 'Hello, how are you?';
 const vowels = ['a', 'e', 'i', 'o', 'u'];
 
 function vowelsInString(sentence){
-	if (sentence.length === 0){ 
-		return 0;
-	} else {
-		if(vowels.includes(sentence[0].toLowerCase())){
-			return 1 + vowelsInString(sentence.slice(1));
-		} else {
-			return vowelsInString(sentence.slice(1));
-		}
-	}
+    if (sentence.length === 0){ 
+        return 0;
+    } else {
+        if(vowels.includes(sentence[0].toLowerCase())){
+            return 1 + vowelsInString(sentence.slice(1));
+        } else {
+            return vowelsInString(sentence.slice(1));
+        }
+    }
 }
 console.log(vowelsInString(sentenceEx4));
+
+
+
+console.log("---Ex4---");
 // 4. Write a recursive function to remove all occurrences of a specified character from a string.
 const sentenceEx5 = 'Hello, how are you?';
 const charToRemove = 'o';
 
+function removeCharacter(sentence, removeWord){
+    if(sentence.length === 0 || sentence[0].length === 0){
+        return "";
+        }
+        if(sentence[0] === removeWord ){
+            return removeCharacter(sentence.slice(1), removeWord)
+        }else {
+            return sentence[0] + removeCharacter(sentence.slice(1), removeWord);
+        }
+}
+
+console.log(removeCharacter(sentenceEx5,charToRemove));
+
+console.log("---Ex5---");
 // 5. Write a recursive function to check if an array includes a specific value.
 const numbers = [1, 2, 3, 4, 5];
-const valueToCheck = 3;
+const valueToCheck = 2;
+
+
+function isArrayIncludeValue(array, value){
+    if(array.length === 0){
+        return "Value is not found";
+    } 
+    if(array[0] !== value){
+        return isArrayIncludeValue(array.slice(1), value);
+    } else{
+        return true;
+    }
+
+}
 
+console.log(isArrayIncludeValue(numbers, valueToCheck));
+console.log("---Ex6---");
 // 6. Write a recursive function to flatten an object with nested objects into a single-level object.
 const nestedObj = {
-	name: 'John',
-	age: 30,
-	address: {
-		street: '123 Main St',
-		city: 'New York',
-		country: 'USA',
-	},
-	hobbies: ['reading', 'coding'],
-	social: {
-		twitter: '@johndoe',
-		linkedIn: 'johndoe',
-	},
+    name: 'John',
+    age: 30,
+    address: {
+        street: '123 Main St',
+        city: 'New York',
+        country: 'USA',
+    },
+    hobbies: ['reading', 'coding'],
+    social: {
+        twitter: '@johndoe',
+        linkedIn: 'johndoe',
+    },
 };
 
+
+
+// I have no idea about this example...
```

Comment:
Here's how to do that:
```
function flattenObject(obj, prefix = '') {  
    let flattened = {}; // initialize the object as an empty object, we will use it later to store intermediate values

    for (let key in obj) {   // iterate through each key in the current object
        if (typeof obj[key] === 'object' && obj[key] !== null) {   // if a value of a key is an object itself, then we need to flatten it
            let nested = flattenObject(obj[key], `${prefix}${key}.`); // call the function recursively and provide the prefix (e.g. address to make it address.street in the next function run)
            flattened = { ...flattened, ...nested }; // update the initially created variable with the result of the operation above
        } else {  
            flattened[`${prefix}${key}`] = obj[key]; // if a key is not an object, save it to `flattened` object and move on to the next key
        }  
    }  
    return flattened;  
}
```

Reply:
I still have problem to understand that. Can you explain it on meet after the lesson.